### PR TITLE
[DataFactory] Event Trigger Subscription APIs

### PR DESCRIPTION
**Pull request by cesher**
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json [hunk 1]
```diff
@@ -2765,6 +2765,162 @@
...
       }
     },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DataFactory/factories/{factoryName}/triggers/{triggerName}/subscribeToEvents": {
+      "post": {
+        "tags": [
+          "triggers"
+        ],
+        "operationId": "Triggers_SubscribeToEvents",
+        "x-ms-examples": {
+          "Triggers_Start": {
```
Comment (hvermis) [inline specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json]: Change the name, must be the same as operation id

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json [hunk 2]
```diff
@@ -2765,6 +2765,162 @@
...
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DataFactory/factories/{factoryName}/triggers/{triggerName}/getEventSubscriptionStatus": {
+      "post": {
+        "tags": [
+          "triggers"
+        ],
+        "operationId": "Triggers_GetEventSubscriptionStatus",
+        "x-ms-examples": {
+          "Triggers_Start": {
```
Comment (hvermis) [inline specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json]: same here

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json [hunk 3]
```diff
@@ -2765,6 +2765,162 @@
...
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "Trigger event subscription state.",
+            "schema": {
+              "$ref": "#/definitions/TriggerSubscriptionOperationStatus"
+            }
+          },
+          "202": {
```
Comment (hvermis) [inline specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json]: Why is this one 202?

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json [hunk 4]
```diff
@@ -2765,6 +2765,162 @@
...
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DataFactory/factories/{factoryName}/triggers/{triggerName}/unsubscribeFromEvents": {
+      "post": {
+        "tags": [
+          "triggers"
+        ],
+        "operationId": "Triggers_UnsubscribeFromEvents",
+        "x-ms-examples": {
+          "Triggers_Start": {
```
Comment (hvermis) [inline specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json]: same here

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json [hunk 5]
```diff
@@ -3658,6 +3814,23 @@
         "message"
       ]
     },
+    "TriggerSubscriptionOperationStatus": {
+      "x-ms-external": true,
```
Comment (hvermis) [inline specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json]: this is not external - remove this line

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json (hunk 3 above)
Comment (cesher) [reply specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json]: fixed it, now it is 200.

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json (hunk 4 above)
Comment (cesher) [reply specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json]: done

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json (hunk 5 above)
Comment (cesher) [reply specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json]: done

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json [hunk 6]
```diff
@@ -2765,6 +2765,156 @@
...
+          "triggers"
+        ],
+        "operationId": "Triggers_GetEventSubscriptionStatus",
+        "x-ms-examples": {
+          "Triggers_GetEventSubscriptionStatus": {
+            "$ref": "./examples/Triggers_GetEventSubscriptionStatus.json"
+          }
+        },
+        "description": "Get a trigger's event subscription status.",
+        "x-ms-long-running-operation": true,
```
Comment (hvermis) [inline specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json]: remove this

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json [hunk 7]
```diff
@@ -3658,6 +3808,22 @@
...
+    "TriggerSubscriptionOperationStatus": {
+      "description": "Defines the response of a trigger subscription operation.",
+      "type": "object",
+      "properties": {
+        "triggerName": {
+          "description": "Trigger name.",
+          "type": "string",
+          "readOnly": true
+        },
+        "status": {
```
Comment (hvermis) [inline specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json]: Do you want to add enum here? what are the possible statuses?

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json (hunk 6 above)
Comment (cesher) [reply specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json]: done

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json (hunk 7 above)
Comment (cesher) [reply specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json]: done

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/examples/Triggers_GetEventSubscriptionStatus.json [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+  "parameters": {
+    "subscriptionId": "12345678-1234-1234-1234-12345678abc",
+    "resourceGroupName": "exampleResourceGroup",
+    "factoryName": "exampleFactoryName",
+    "triggerName": "exampleTrigger",
+    "api-version": "2018-06-01"
+  },
+  "responses": {
+    "200": {
+      "headers": {
```
Comment (hvermis) [inline specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/examples/Triggers_GetEventSubscriptionStatus.json]: this shouldn't return location header

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json [hunk 9]
```diff
@@ -3658,6 +3807,40 @@
...
+          "description": "Subscription status.",
+          "$ref": "#/definitions/EventSubscriptionStatusTypes",
+          "readOnly": true
+        }
+      }
+    },
+    "EventSubscriptionStatusTypes": {
```
Comment (hvermis) [inline specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json]: You don't need to declare it a new type. The types make sense when they are used or can be used in future in multiple APIs.  It doesn't seem to be the case here. Also the type you have here is array - can you returned object have multiple statuses?
You should just do something similar to this TriggerRunStatus for your status property. And remove types from the name.
       "status": {
          "type": "string",
          "enum": [
            "Succeeded",
            "Failed",
            "Inprogress"
          ],
          "x-ms-enum": {
            "name": "TriggerRunStatus",
            "modelAsString": true
          },
          "description": "Trigger run status.",
          "readOnly": true

**Review by hvermis — APPROVED** [review]
Please address last 2 comments

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/examples/Triggers_GetEventSubscriptionStatus.json (hunk 8 above)
Comment (cesher) [reply specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/examples/Triggers_GetEventSubscriptionStatus.json]: done

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json [hunk 10]
```diff
@@ -3658,6 +3807,51 @@
...
+            "name": "EventSubscriptionStatus",
+            "modelAsString": true
+          },
+          "description": "Event Subscription Status.",
+          "readOnly": true
+        }
+      }
+    },
+    "EventSubscriptionStatusTypes": {
+      "type": "array",
```
Comment (hvermis) [inline specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json]: you didn't remove this?

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json [hunk 11]
```diff
@@ -2765,6 +2765,155 @@
...
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "Trigger is subscribed to events.",
+            "schema": {
+              "$ref": "#/definitions/TriggerSubscriptionOperationStatus"
+            }
+          },
+          "202": {
```
Comment (hvermis) [inline specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json]: Missed this one - 202 doesn't return status object, just Accepted code:
          "202": {
            "description": "Accepted."
          },

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json [hunk 12]
```diff
@@ -2765,6 +2765,155 @@
...
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "Trigger is unsubscribed from events.",
+            "schema": {
+              "$ref": "#/definitions/TriggerSubscriptionOperationStatus"
+            }
+          },
+          "202": {
```
Comment (hvermis) [inline specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json]: same here:
          "202": {
            "description": "Accepted."
          },

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json (hunk 11 above)
Comment (cesher) [reply specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json]: done

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json (hunk 12 above)
Comment (cesher) [reply specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json]: done

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json (hunk 10 above)
Comment (cesher) [reply specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/datafactory.json]: removed

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/examples/Triggers_SubscribeToEvents.json [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+    "factoryName": "exampleFactoryName",
+    "triggerName": "exampleTrigger",
+    "api-version": "2018-06-01"
+  },
+  "responses": {
+    "202": {
+      "headers": {
+        "Location": "https://management.azure.com/subscriptions/12345678-1234-1234-1234-12345678abc/resourceGroups/exampleResourceGroup/providers/Microsoft.DataFactory/factories/exampleFactoryName/triggers/exampleTrigger/getEventSubscriptionStatus?api-version=2018-06-01"
+      },
+      "body": {
```
Comment (hvermis) [inline specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/examples/Triggers_SubscribeToEvents.json]: delete this body too

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/examples/Triggers_UnsubscribeFromEvents.json [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+    "factoryName": "exampleFactoryName",
+    "triggerName": "exampleTrigger",
+    "api-version": "2018-06-01"
+  },
+  "responses": {
+    "202": {
+      "headers": {
+        "Location": "https://management.azure.com/subscriptions/12345678-1234-1234-1234-12345678abc/resourceGroups/exampleResourceGroup/providers/Microsoft.DataFactory/factories/exampleFactoryName/triggers/exampleTrigger/getEventSubscriptionStatus?api-version=2018-06-01"
+      },
+      "body": {
```
Comment (hvermis) [inline specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/examples/Triggers_UnsubscribeFromEvents.json]: same here - delete

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/examples/Triggers_SubscribeToEvents.json (hunk 13 above)
Comment (cesher) [reply specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/examples/Triggers_SubscribeToEvents.json]: done

File: specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/examples/Triggers_UnsubscribeFromEvents.json (hunk 14 above)
Comment (cesher) [reply specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/examples/Triggers_UnsubscribeFromEvents.json]: done

**Comment by jhendrixMSFT** [comment]
Looks good from SDK perspective, waiting for ARM to sign off.

**Comment by hvermis** [comment]
@jhendrixMSFT Who is on call from ARM PMs this week?

**Comment by jhendrixMSFT** [comment]
@ryansbenson is the ARM on-call this week

**Comment by ravbhatnagar** [comment]
Signing off from ARM side.

**Review by jhendrixMSFT — APPROVED** [review]

**Pull request merged**